temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java

```diff
@@ -100,21 +155,160 @@ public <T> T fromPayloads(
+    return this.encodeFailure(dataConverter.exceptionToFailure(e).toBuilder()).build();
+  }
+
+  private Failure.Builder encodeFailure(Failure.Builder failure) {
+    if (failure.hasCause()) {
+      failure.setCause(encodeFailure(failure.getCause().toBuilder()));
+    }
+    if (this.encodeDefaultAttributes) {
+      Map<String, String> encodedAttributes = new HashMap<>();
+      encodedAttributes.put(STACK_TRACE_KEY, failure.getStackTrace());
+      encodedAttributes.put(MESSAGE_KEY, failure.getMessage());
+      Payload encodedAttributesPayload = toPayload(Optional.of(encodedAttributes)).get();
+      failure
+          .setEncodedAttributes(encodedAttributesPayload)
+          .setMessage(ENCODED_FAILURE_MESSAGE)
+          .setStackTrace("");
+    }
+    switch (failure.getFailureInfoCase()) {
+      case APPLICATION_FAILURE_INFO:
+        {
+          ApplicationFailureInfo.Builder info = failure.getApplicationFailureInfo().toBuilder();
+          if (info.hasDetails()) {
+            info.setDetails(encodePayloads(info.getDetails()));
+          }
+          failure.setApplicationFailureInfo(info);
+        }
+        break;
+      case TIMEOUT_FAILURE_INFO:
+        {
+          TimeoutFailureInfo.Builder info = failure.getTimeoutFailureInfo().toBuilder();
+          if (info.hasLastHeartbeatDetails()) {
+            info.setLastHeartbeatDetails(encodePayloads(info.getLastHeartbeatDetails()));
+          }
+          failure.setTimeoutFailureInfo(info);
+        }
+        break;
+      case CANCELED_FAILURE_INFO:
+        {
+          CanceledFailureInfo.Builder info = failure.getCanceledFailureInfo().toBuilder();
+          if (info.hasDetails()) {
+            info.setDetails(encodePayloads(info.getDetails()));
+          }
+          failure.setCanceledFailureInfo(info);
+        }
+        break;
+      case RESET_WORKFLOW_FAILURE_INFO:
+        {
+          ResetWorkflowFailureInfo.Builder info = failure.getResetWorkflowFailureInfo().toBuilder();
+          if (info.hasLastHeartbeatDetails()) {
+            info.setLastHeartbeatDetails(encodePayloads(info.getLastHeartbeatDetails()));
+          }
+          failure.setResetWorkflowFailureInfo(info);
+        }
+        break;
+      default:
+        {
+          // Other type of failure info don't have anything to encode
+        }
+    }
+    return failure;
+  }
+
+  @Override
+  @Nonnull
+  public RuntimeException failureToException(@Nonnull Failure failure) {
+    Preconditions.checkNotNull(failure, "failure");
+    return dataConverter.failureToException(this.decodeFailure(failure.toBuilder()).build());
+  }
+
+  private Failure.Builder decodeFailure(Failure.Builder failure) {
+    if (failure.hasCause()) {
+      failure.setCause(decodeFailure(failure.getCause().toBuilder()));
+    }
+    if (failure.hasEncodedAttributes()) {
+      Map<String, String> encodedAttributes =
+          fromPayload(failure.getEncodedAttributes(), HashMap.class, HASH_MAP_STRING_STRING_TYPE);
```

Comment:
I think you should use a static inner pojo that represents this attribute as a Jackson type for clearer conversion and field naming.

(restating previous comment since it was outdated)